core/controllers/ingest.py

```diff
@@ -59,63 +63,194 @@ def ingest(body: dict, excel_file: FileStorage) -> tuple[dict, int]:
-        "loaded": do_load,
-    }
+        load_data(transformed_data, excel_file)
+
+    programme_metadata = get_metadata(transformed_data)
+    return build_success_response(programme_metadata=programme_metadata, do_load=do_load)
 
-    return success_payload, 200
 
+def parse_body(body: dict) -> tuple[str, int, dict | None, bool]:
+    """Parses the request body.
 
-def process_initial_validation_errors(error_messages: list[str]) -> tuple[dict, int]:
+    :param body: request body
+    :return: parsed values
+    """
+    fund = body.get("fund_name")
+    reporting_round = body.get("reporting_round")
+    auth = parse_auth(body)
+    do_load = body.get("do_load", True)  # defaults to True, if False then do not load to database
+    return fund, reporting_round, auth, do_load
+
+
+def extract_process_validate_tables(
+    workbook_data: dict[str, pd.DataFrame], tables_config: dict[str, dict]
+) -> dict[str, Table]:
+    """Extracts, processes and validates tables from a workbook based on the specified configuration.
+
+    :param workbook_data: a dictionary containing worksheet names as keys and corresponding pandas DataFrames as values
+    :param tables_config: a dictionary containing table names as keys and corresponding configuration dictionaries as
+        values
+    :return: a dictionary containing table names as keys and lists of extracted tables as pandas DataFrames as values
+    :raises ValidationError: if the data fails validation
+    """
+    extractor = ta.TableExtractor(workbook_data)
+    tables = {}
+    error_messages = []
+    for table_name, config in tables_config.items():
+        worksheet_name = config["extract"]["worksheet_name"]
+        extracted_tables = extractor.extract(**config["extract"])
+        processor = ta.TableProcessor(**config["process"])
+        validator = ta.TableValidator(config["validate"])
+        # All PFV1 tables are singular, so we assume there is only one table. This may not be true for future templates.
+        table = extracted_tables[0]
+        processor.process(table)
+        tables[table_name] = table
+        try:
+            validator.validate(table)
+        except ta.TableValidationErrors as e:
+            for error in e.validation_errors:
+                error_messages.append(
+                    Message(
+                        sheet=worksheet_name,
+                        section=None,
+                        cell_index=error.cell.str_ref if error.cell else None,
+                        description=error.message,
+                        error_type=None,
+                    )
+                )
+                current_app.logger.info(
+                    f"{config['extract']['worksheet_name']} {error.cell.str_ref if error.cell else ''}:"
+                    f" {error.message}"
+                )
+    if error_messages:
+        raise ValidationError(error_messages)
+    return tables
+
+
+def coerce_data(data: dict[str, pd.DataFrame], config: dict):
```

Comment:
The issue is that we currently rely on custom `Check`s (e.g. `is_float`) to determine if datatypes are correct. We also force `coerce` to be `False` because if we did allow `coerce` to be `True` then we'd need to decide how to handle the `coerce_dtype` errors that could then be raised. 

We could ignore those failures, like we do `dtype` errors, but that could then lead to incorrect types going unnoticed if we miss out a type `Check` or if one doesn't function as expected.

If we do the coerce step separately then we can first rely on the type `Check`s defined in the schema with user error messages. If it then fails when we coerce then we know we've made a programming error somewhere.

Reply:
I think a good solution for now is to move the coercion to be done after the schema validation, but before the cross-table validation. Because the results of the cross-table validation shouldn't effect the coercion anyway.